clang_format.py

```diff
@@ -76,22 +120,29 @@ def check_hash(sha: str, file: Path) -> None:
 
     if d.hexdigest() != sha:
         print(
-            f"FATAL: sha1sum mismatch on {file}.  Expected {sha}, but was {d.hexdigest()}",
+            f"FATAL: sha1sum mismatch on {file}.\
+Expected {sha}, but was {d.hexdigest()}",
             file=sys.stderr,
         )
         print("Maybe the file is corrupted?  Try deleting it.")
         sys.exit(1)
 
 
-def clang_format_path() -> Path:
-    """Gets the path of the relevant clang-format binary.
+def get_version_key(version: str) -> Tuple[int, int, int]:
+    major, minor, patch = version.split(".")
+    v = (int(major), int(minor), int(patch))
+    return v
+
 
+def clang_format_path(version: Tuple[int, int, int]) -> Path:
+    """
+    Gets the path of the relevant clang-format binary.
     Downloads it if necessary.
     """
     try:
         base_cachedir = Path(os.environ["XDG_CACHE_HOME"])
```

Comment:
A common pattern when using `pre-commit` in the CI is the following:

- Override `PRE_COMMIT_HOME = cache/pre-commit`
- Tell the CI system to cache the `PRE_COMMIT_HOME` path
- Then we don't initialise the virtual environments multiple times

This doesn't work well with that because the cache for the clang format binary is separate to the location of the initialised virtual environment.

I would recommend downloading to a local cache director `Path(__file__).parent / ".cache"`:

- `pre-commit` automatically attempts to deduplicate pre-commit hooks in the shared `PRE_COMMIT_HOME` location
- we would naturally work with normal suggested caching of `pre-commit` hooks on CI systems
- It is expected that once a hook is "initialized" it is good to go

Or, if not that, at least default to `PRE_COMMIT_HOME` first.

Reply:
This isn't needed for this patch. It just reminded me when I saw the `Path.home()` change. I can submit a separate patch for this.